Alternate escape shuttles and ferries

coiax opened the pull request:
It's quite common  for admins to spawn drone shells on Centcom or just build mode and make customised shuttles for events or giggles.

So it would be good if we had a pool of customised shuttles to pull from.

I have the idea that there would be some adminonly computer on Centcom that would replace the shuttle with a different one, kind of like how the holodeck copies its contents from special locked boxes on the Central z-level.
- [x] make a bunch of escape shuttles
- [x] code a special "invulnerable" subdatum of drone
- [x] determine how the brig on the shuttle works for the "escape not in custody component"
- [x] make a bunch of centcom ferries
- [x] work out an easy way for admins to swap them in and out, maybe prepare some centcom private/public messages

:cl: coiax
rscadd: To save credits, Centcom has subcontracted out emergency shuttle design to third parties. Rest assured, all of these produced shuttles meet our stringent quality control.
/:cl:

Current shuttles are:
# The Emergency Escape Bar

Features include a very tough Bardrone (3000hp because I couldn't find the godmode flag), with laws to be a good bartender, bathroom (with a locker of janitor supplies for those accidental spills), a quality lounge for the heads, and a small gambling table.

![Emergency Escape Bar](http://i.imgur.com/yWCZz2d.png)
# Secure Transport Vessel 5 (STV5)

Well, looks like Centcomm only had this ship in the area, they probably weren't expecting you to need evac for a while. Probably best if you don't rifle around in whatever equipment they were transporting. I hope you're friendly with your coworkers, because there is very little space in this thing.

Contains contraband armory guns, maintenance loot, and abandoned crates!

![Secure Transport Vessel 5](http://i.imgur.com/M2EV3Gl.png)
# Shuttle 667

Looks like this shuttle may have wandered into the darkness between the stars on route to the station. Let's not think too hard about where all the bodies came from.

![Shuttle 667](http://i.imgur.com/i9dkD1d.png)
# Hyperfractal Gigashuttle

"I dunno, this seems kinda needlessly complicated."
"This shuttle has very a very high safety record, according to Centcom
Officer Cadet Yins."
"Are you sure?"
"Yes, it has a safety record of N-A-N, which is apparently larger than
100%."

![Hyperfractal Gigashuttle](https://cloud.githubusercontent.com/assets/609465/14925751/e28485fa-0e40-11e6-963d-9050f81d99ec.png)
# "Meat" Ferry

Ahoy! We got all kinds o' meat aft here. Meat from plant people, people who be dark, not in a racist way, just they're dark black. Oh and lizard meat too,mighty popular that is. Definitely 100% fresh, just ask this guy here. _person on meatspike moans_ See? Definitely high quality meat, nothin' wrong with it, nothin' added, definitely no zombifyin' reagents! Why be you talkin' about zombies anyway, I'll sue you, arr!
!["Meat" Ferry](https://cloud.githubusercontent.com/assets/609465/14683977/7aa08c30-0725-11e6-9895-c028b446d4c8.png)
# The Lighthouse(?)

_static_... part of a much larger vessel, possibly military in origin.
The weapon markings aren't anything we've seen ... _static_ ... by
almost never the same person twice, possible use of unknown storage ...
_static_ ... seeing ERT officers onboard, but no missions are on file
for ... _static_ ... _static_ ... _annoying jingle_ ... only at The
LIGHTHOUSE! Fulfilling needs you didn't even know you had. We've got
EVERYTHING, and something else!
![The Lighthouse](https://cloud.githubusercontent.com/assets/609465/14720412/43b36678-07f9-11e6-892b-bfaabdb45dc5.png)
# Snappop(tm)!

Hey kids and grownups! Are you bored of DULL and TEDIOUS shuttle journeys after you're evacuating for probably BORING reasons. Well then order the Snappop today! We've got fun activities for everyone, an all access cockpit, and no boring security brig! Boo! Play dress up with your friends! Collect all the bedsheets before your neighbour does so! Check if the AI is watching you with our patent pending "Peeping Tom AI Multitool Detector"  or PEEEEEETUR for short. Have a fun ride!
![Snappop(tm)!](https://cloud.githubusercontent.com/assets/609465/14909138/d4d3f134-0dd7-11e6-8459-3499c3982e9f.png)

coiax committed:
Added a simple bar-theme escape shuttle

Features include a very tough Bardrone, with laws to be a good
bartender, bathroom, a quality lounge for the heads, and a small
gambling table.

coiax committed:
Changed drone's health to 3000

Bardrone is now strong, to cope from stray gunfire.

coiax committed:
Replaces banned variables with alternates

Step_[xy] is bad, mkay?

coiax committed:
Added STV5, the cramped transport shuttle

Well, looks like Centcom only had this ship ready for evacuation. I hope
you're friendly with your coworkers!

coiax committed:
Removed shuttle area from empty space on STV5

coiax committed:
Added cult shuttle

Looks like this automated shuttle may have wandered into the darkness
between the stars on route to the station. Let's not think too hard
about where all the bodies came from.

swankcookie commented:
very :+1: .  2 questions)  1: what is in those cloners? is that a cult specific sprite? and 2: Maybe making the shuttle kind of dangerous, and not just a safe haven for the crew, would prevent the epidemic of early shuttle calls...

coiax commented:
When a cloner is EMP'd while cloning someone, there's a chance of things going HORRIBLY WRONG. That's the sprite that I set it to.

The thing about making the escape shuttle dangerous is that sometimes the crew do just desperately need to escape. But making a collection of "mean" escape shuttles for the admins to use if they feel the crew are being cowardly or lazy, that would be useful.

swankcookie commented:
I totally agree

coiax committed:
Include the Narnar Shuttle

Dorsisdwarf commented:
The cramped one is fun if there's a big crew and the captain red-alerts a "we bored lets go home"

PKPenguin321 commented:
hail skub

RemieRichards commented on tgstation.dme:
```diff
@@ -15,6 +15,9 @@
 // BEGIN_INCLUDE
 #include "_maps\__MAP_DEFINES.dm"
 #include "_maps\tgstation2.dm"
+#include "_maps\barshuttle.dmm"
```
no.. no, no you don't do this.

coiax replied:
I assume there's probably a better place to put the map files. But I needed to put them somewhere.

RemieRichards replied:
No it's the fact that you don't put these maps in the DME, at all.

coiax replied:
oh, I see. Sorry, I think Dream Maker changed it automatically, and I must have committed by accident.

RemieRichards replied:
well, commit an update to remove them from the DME and it'll be fine.

AdamElTablawy commented:
>gun filled escape shuttle

wow this won't go horribly wrong at all

coiax commented:
> > gun filled escape shuttle
> wow this won't go horribly wrong at all

There are two armoury contraband guns in Security ID locked crates. wouldn't call that "filled" tbh

AdamElTablawy commented:
someone's still gonna run on, emag it, and grab some extra ammunition to whack off their murderboner with

also the shuttle's so small it's gonna get welderbombed every time

AdamElTablawy commented:
also make bardrone unable to be picked up-if you're giving it 3000 health it's obviously not intended to die too easily, so it being picked up and spaced is anticlimactic

Jordie0608 commented:
>  determine how the brig on the shuttle works for the "escape not in custody component"

https://github.com/tgstation/-tg-station/blob/master/code/game/gamemodes/objective.dm#L351

coiax commented:
> also make bardrone unable to be picked up-if you're giving it 3000 health it's obviously not intended to die too easily, so it being picked up and spaced is anticlimactic

I might just make a special "invulnerable" drone datum child that's got all of these things set up, instead of fiddling around with map maker variables.

coiax commented:
@Jordie0608 thanks.

optimumtact commented:
FIND THE FLIPPING GODMODE VAR AEIOU

optimumtact commented:
@AdamElTablawy given these are admin only I can't think of a reason to be very concerned about what's on them.

If the admins are touching the round, shit's gonna be unbalanced anyway.

Cheridan commented:
Should just make the drone a serviceborg

coiax commented:
Service borgs can only dispense drinks from their own internal storage, making the whole bar setup redundant. Also, someone will probably just jump the bar and emag it at some point, and that's just not cool.

TrustyGun commented:
why not just leave out the drone and make the vender available for everyone

also, you may need to make some variants of these for meta

AdamElTablawy commented:
the drone needs a custom lawset so it isn't breaking it's laws to vend drinks to humans

coiax commented:
Drones are cool. And it does have a custom lawset.

Right now I'm just working out how to make a way of having admins easily be able to hotswap these things in and out, lowering the barrier to entry.

coiax committed:
Meat ferry added

Very basic, could probably use some love from a map wizard.

coiax committed:
Merge branch 'master' into alternate-escape-shuttles

Jalleo commented:
Image for meat ferry?

swankcookie commented:
...what?

Dorsisdwarf commented:
metastation, I assume

coiax commented:
Meat ferry was just an attempt to make a meat trader ferry in the ferry slot. Dunno if it's any good.

coiax committed:
Merge branch 'master' into alternate-escape-shuttles

coiax committed:
Moved shuttle files to shuttles/ directory

coiax committed:
MAPS MOVED AND RENAMED

coiax committed:
Hyperfractal Gigashuttle

"I dunno, this seems kinda needlessly complicated."
"This shuttle has very a very high safety record, according to Centcom
Officer Cadet Yins."
"Are you sure?"
"Yes, it has a safety record of N-A-N, which is apparently larger than
100%."

coiax committed:
I REGRET ADDING THESE

coiax committed:
Into templates you go

coiax committed:
Changed floors

TrustyGun commented:
lmao that hyperfractal gigashittle should be used for the capt attempting to call during blob/rev

:+1:

coiax committed:
Hugbox supermatter, for map making

coiax committed:
Shuttles KINDA delete themselves

coiax committed:
Hats in hats in hats in hats in hats

coiax committed:
Merge branch 'master' into alternate-escape-shuttles

coiax committed:
Use the hugbox supermatter in the shuttle

coiax committed:
Include the new verbs

coiax committed:
Include our new verbs

coiax committed:
Fixed runtime when deleting emergency shuttle

coiax committed:
Added to verb list

coiax committed:
Fixes more runtimes with deleting the shuttle

coiax committed:
Shuttle Destroy verb now functions

coiax committed:
Moved them out of the shuttles/ folder

coiax committed:
We'll need these docks later

coiax committed:
shuttle_import landmark datum made

coiax committed:
Changes z2 to TGM format and adds shuttle_import landmark

coiax committed:
SHUTTLE IMPORT VERB

Is this it? Are we there?

coiax committed:
Some compile fixes

Dorsisdwarf commented:
Does the hyperfractal explode the -ooh verb- shuttle? 3 supermatters being constantly fired at by emitters sounds like !FUN!

coiax commented:
Nah, these are special HUGBOX supermatter that don't emit gas or get damaged. They still irradiate, hallucinate anything nearby. And obviously full DUST FUN.

Anyway, we'll squash all the commits when it comes for testing merging.

Dorsisdwarf commented:
Might as well make it a singulo, honk

coiax committed:
Emergency bar map updated

coiax committed:
Updated cramped shuttle

coiax commented:
Well, problem with singulo is that unless you put it on some singularity proof plating, it'll just fall out of the shuttle as soon as it starts moving.

coiax commented:
Now a TESLA, that's another story.

Dorsisdwarf commented:
Oh yeah, singulo can't hop aboard HMS Doomed now

coiax committed:
Updated narnar shuttle

coiax committed:
Supermatter shuttle updated

coiax committed:
Meat ferry updated

coiax committed:
Debugging information included

coiax committed:
Debug code, QDEL hints

coiax committed:
Minor updates to collection of maps

coiax committed:
May as well use the snowflake drone code

PICK YOUR OWN BARKEEPING HAT.

coiax committed:
The Lighthouse

*static*... part of a much larger vessel, possibly military in origin.
The weapon markings aren't anything we've seen ... *static* ... by
almost never the same person twice, possible use of unknown storage ...
*static* ... seeing ERT officers onboard, but no missions are on file
for ... *static* ... *static* ... *annoying jingle* ... only at The
LIGHTHOUSE! Fulfilling needs you didn't even know you had. We've got
EVERYTHING, and something else!

ChangelingRain commented on code/modules/admin/verbs/shuttles.dm:
```diff
@@ -0,0 +1,91 @@
+		return
+
+	var/datum/map_template/template
+	// TODO would be helpful if it only highlighted actual shuttles rather
```
TODO?

AnturK replied:
Probably should be done so loading some lavaland crashed shuttle or similar won't fool the admins

coiax replied:
If the admins go into the LAVALANDTEMPLATES folder, then that's on them. Anyway, if they accidentally import a lavaland crashed shuttle template, that'll just bring it into existence in the shuttle import section in space. No negative effects apart from the random template on z2.

RandomMarine commented:
GET_DOCKED

coiax committed:
Made supermatter hallucination too powerful

coiax committed:
Added force argument for docking testing

MrStonedOne commented:
So, the take away from irc, is this will need more work to be more safe, spawning some fail safe shuttle, rather then adding a bunch of if checks to ticker code. I'd like ticker code to be able to assume the shuttle exists.

coiax committed:
Added emergency backup shuttle

Hopefully it will never be seen or used. But means that provided it's
there and untouched, even with the accidental deletion of a shuttle,
round will still be able to end as normal.

coiax committed:
Added templates for existing shuttles

coiax committed:
Merge branch 'master' into alternate-escape-shuttles

Conflicts:
	_maps/map_files/generic/z2.dmm

coiax committed:
Backup shuttle used if emergency shuttle deleted

coiax committed:
Added backup shuttle transit space

Don't know why transit spaces aren't dynamically generated, to be
honest.

coiax committed:
Hyperfractal Gigashuttle improved

More radiation lockers, chairs which makes it "safer".

coiax committed:
Added Shuttle Under Construction

The documentation hasn't been finished yet for this shuttle. Break glass
in case of emergency.

coiax committed:
Added warnings for shuttle interaction

ARE YOU SURE YOU WANT TO DELETE THE EMERGENCY SHUTTLE?

I mean, I really don't know what will happen, and it's probably
nothing good.

coiax committed:
Merge branch 'master' into alternate-escape-shuttles

coiax committed:
Snappop(tm)!

Hey kids and grownups! Are you bored of DULL and TEDIOUS shuttle
journeys after you're evacuating for probably BORING reasons. Well then
order the Snappop today! We've got fun activities for everyone, an all
access cockpit, and no boring security brig! Boo! Play dress up with
your friends! Collect all the bedsheets before your neighbour does so!
Check if the AI is watching you with our patent pending "Peeping Tom AI
Multitool Detector" or PEEEEEETUR for short. Have a fun ride!

Razharas commented:
@coiax say when its ready

coiax commented:
@Razharas it's ready in my opinion.

coiax committed:
Fixed horrific experiment icon, added gambling spawner

coiax committed:
Candle changes

Infinite candles start fires but don't run out.
Infinite candles don't make messages when created.

coiax committed:
Shuttle can dock if already docked

coiax committed:
Merge branch 'alternate-escape-shuttles' of github.com:coiax/-tg-station into alternate-escape-shuttles

Jordie0608 commented:
\<Jordie\> looks alright coiax but im tired so who knows ¯\\_(ツ)_/¯

coiax committed:
Fixing bugs with shuttle import

Also, some minor bugs with hugbox supermatter still consuming
gas moles and slowly depressurising the area.

Jordie0608 commented on code/modules/admin/verbs/shuttles.dm:
```diff
@@ -0,0 +1,108 @@
+		M = AM
+		names += M.name
+
+	var/selected = input("Select shuttle to DESTROY", "Shuttles") in names
```
There should really be a cancel button on this input()

coiax replied:
I don't know if there's a way of adding cancel buttons to basic byond input(). It does ask afterwards. Obviously when this is all replaced by the shuttle manipulator tgui, then it won't matter.

tigercat2000 replied:
input("name", "title") as null|anything in list

coiax replied:
brilliant, thanks.

Jordie0608 commented:
When trying to import ferry_lighthouse.dmm

> runtime error: 
[15:34:47]Cannot read null.x
[15:34:47]proc name: load (/datum/map_template/proc/load)
[15:34:47]  source file: map_template.dm,27
[15:34:47]  usr: Pip (/mob/dead/observer)
[15:34:47]  src: ferry_meat.dmm (/datum/map_template)
[15:34:47]  usr.loc: space (119,113,2) (/turf/open/space)
[15:34:47]  call stack:
[15:34:47]ferry_meat.dmm (/datum/map_template): load(null, 1)
[15:34:47]Jordie0608 (/client): Shuttle Import()
runtime error: 
[15:34:47]Cannot read null.x
[15:34:47]proc name: get affected turfs (/datum/map_template/proc/get_affected_turfs)
[15:34:47]  source file: map_template.dm,73
[15:34:47]  usr: Pip (/mob/dead/observer)
[15:34:47]  src: ferry_meat.dmm (/datum/map_template)
[15:34:47]  usr.loc: space (119,113,2) (/turf/open/space)
[15:34:47]  call stack:
[15:34:47]ferry_meat.dmm (/datum/map_template): get affected turfs(null, 1)
[15:34:47]Jordie0608 (/client): Shuttle Import()
The loaded template didn't have a mobile docking port!
If you know what you're doing, you can make your own. But this verb cannot help you past that point.

Jordie0608 commented on code/modules/admin/verbs/shuttles.dm:
```diff
@@ -0,0 +1,108 @@
+
+	if(!M)
+		usr << "<span class='warning'>The loaded template didn't have a mobile docking port!</span>"
+		// TODO maybe unload the template? Pff, let the admins clean up
```
=\

Jordie0608 commented on code/modules/admin/verbs/shuttles.dm:
```diff
@@ -0,0 +1,108 @@
+	var/datum/map_template/template
+	// TODO would be helpful if it only highlighted actual shuttles rather
+	// than ALL OF THEM
+	var/map = input(usr, "Choose a Shuttle Template to import","Import Shuttle Template") as null|anything in map_templates
```
Why not just populate a list of only shuttle templates like how space and lava ruins ones do?

Jordie0608 commented on code/modules/admin/verbs/shuttles.dm:
```diff
@@ -0,0 +1,108 @@
+
+	var/selected = input("Select shuttle to DESTROY", "Shuttles") in names
+
+	var/decide_against_msg = "You decide against destroying a shuttle."
```
Kinda unnecessary

coiax replied:
I liked giving output when aborting the decision, so you could know that it didn't error out.

Jordie0608 commented on code/modules/shuttle/shuttle.dm:
```diff
@@ -18,9 +18,18 @@
+	else
+		// If not removed immediately, it can interfere with the docking
+		// detection code, which is annoying and inconvinient.
+		return QDEL_HINT_HARDDEL_NOW
```
What sort of things prevent this from having its references cleared and qdelling? Is it just that it needs to happen real quicksmart?

coiax replied:
The docking port was lingering long enough to interfere with later attempts to import new shuttles, because the docking code was picking up the existence of the doomed docking port and refusing to let the imported shuttle dock.

Jordie0608 commented:
This happened when importing emergency_cramped.dmm, emergency_clown.dmm has been imported then deleted prior to this.
![space station 13 2016-05-01 155417](https://cloud.githubusercontent.com/assets/4343468/14940314/03a2004e-0fb5-11e6-8ce1-97d0e5aa28d9.png)

Jordie0608 commented on code/game/gamemodes/game_mode.dm:
```diff
@@ -172,7 +172,8 @@
 		return replacementmode.check_finished()
-	if(SSshuttle.emergency.mode >= SHUTTLE_ENDGAME || station_was_nuked)
+	if((SSshuttle.emergency && SSshuttle.emergency.mode >= SHUTTLE_ENDGAME) \
+		 || station_was_nuked)
```
Whats with this strange formatting?

coiax replied:
Avoiding lines of over 79 characters. Force of habit.

optimumtact replied:
well you could always calculate some subvars

    var/emergencymode = SSshuttle.emergency && SSshuttle.emergency.mode
    if(emergencymode || station_was_nuked)

anyway either way works

Jordie0608 commented on code/modules/admin/verbs/shuttles.dm:
```diff
@@ -0,0 +1,108 @@
+		src << "<span class='warning'>Something went wrong, the selected shuttle doesn't exist anymore."
+		return
+
+	log_admin("ShuttleDestroy: [M]")
```
Might as well log the admin here as well

Jordie0608 commented on code/modules/admin/verbs/shuttles.dm:
```diff
@@ -0,0 +1,108 @@
+
+	M.dockRoundstart()
+
+	log_admin("ShuttleImport: [map] - [M]")
```
Might as well log the admin here as well

Jordie0608 commented:
When importing a shuttle, a check that one of the type you've selected doesn't already exist would be good.

coiax commented:
So with your shuttle_cramped import, looks like it's brought part of the cargo shuttle with it, don't know why.

Checking the type of the imported shuttle can only be done after the template's been loaded because all the metadata, as it were, is inside the mobile docking port object.

coiax committed:
Admin notice, unload bad template

coiax committed:
Fixes problems with some shuttles

coiax committed:
Warning message if roundstart dock failed

coiax committed:
Compiles.

coiax committed:
Inform admins of shuttle problems

coiax committed:
Names for shuttles, minor tweaks

peoplearestrange commented:
My issue is that a lot of these designs take away from the basic requirements of the evac shuttle (A brig, a Bridge, a medical bay). Often when admins re-build the shuttle, these 3 places are left un-touched, effecting only the "seating area" part. I'd say the brig is particularly important otherwise we'll see more and more executions of criminals & known traitors simply because there is no way to cart them back to central.

Also, honestly,  you're kinda taking away our fun when the rounds are slow/too crazy to add an admin touch anywhere else. It also makes an "unusual" shuttle a rarity rather than the norm', with everyone putting their own flair to the shuttle, kinda making this automation feel become stale.

**Also if shuttles are changed/picked upon map load your inhibiting any admin who wants to use either shuttle for their own event.**

coiax commented:
The shuttle import are admin only, and aren't automatic, and I wouldn't say being "functional" is top in their design.

The original shuttle designs are in the list if minds are changed.

The PR is mainly about the system for allowing shuttle imports. The shuttles themselves are my shitty shuttle ideas to give some meat to the utensils.

peoplearestrange commented:
Right fair enough! Wasn't understanding. Now am's :3 :+1:

coiax committed:
Merge branch 'master' into alternate-escape-shuttles

coiax committed:
Moved shuttle templates to shuttles/

coiax committed:
Version numbers are lame

coiax committed:
Shuttle datums

coiax committed:
Shuttle manipulator barebones

It doesn't do anything at the moment.

coiax committed:
Added shuttle manipulator to Z2

It still doesn't do anything yet, but it looks pretty.

coiax committed:
Use GetAllContents(), don't delete ghosts

coiax commented:
No merge yet, found a bug that needs fixing.

coiax committed:
Fixes bug where nothing would actually load

coiax commented:
Okay, we're good.

coiax committed:
Cancel button on Destroy Shuttle

coiax committed:
Fixed birdboat shuttle overwriting bar shuttle
